### PR TITLE
feat: mark single notification as read

### PR DESCRIPTION
**Pull request by Leona-cloud**
## What does the PR do?

This feature allows users to mark a specific notification as read. The endpoint takes a notification ID as a parameter and updates the is_read status of the notification.

How should this be manually tested?

1. Login to a user account, and get the access token

2. Make a PATCH request to the `\api\v1\notifications\{notification_id}` endpoint with the following payload:

```json
Authorization: Bearer <access_token>

{
   " is_read": true
}
```

Successful Response
On a successful notification update, the API should return a 200 status code.

```json
{
      "status": "success",
     "status_code": 200,
     "data": {
         "notification_id": "uuid",
         "is_read": true,
        "updated_at": "2024-07-28T12:00:00Z"
      }
}
```

Authentication Error:
When the current user is not properly authenticated to change the organization detail:

```json
{
  "status": "error",
  "message": "Unauthorized request",
  "status_code": 401
}
```

Invalid UserID Error
On Invalid error, if the user is not found with the provided id:
```json
{
  "status": "error",
  "message": "Notification not found",
  "status_code": 404
}
```

Validation:
If any required field is missing or invalid, the response should return a 400 Bad Request status code with appropriate validation error messages.

```json
{
  "status": "error",
  "message": ["notification_id must be a string"],
  "status_code": "400"
}
```
Task

- [x] Implemented PATCH /api/v1/notifications/{notification_id} endpoint in notifications.controller.ts.
- [x] Added method to update the read status of a notification in the NotificationsService.
- [x] Tested endpoint functionality to ensure it handles various scenarios correctly.
- [x] Added appropriate HTTP status codes and response messages for edge cases.
- [x] Added unit tests for the NotificationsService.
- [x] Updated API documentation to reflect the new endpoint.

Reference Issue: [Link to Issue](https://linear.app/team-bingo/issue/BAC2-139/clearing-a-single-or-multiple-notification)

### CODE REVIEW

File: src/modules/notifications/notifications.controller.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import { ApiBearerAuth, ApiBody, ApiOperation, ApiResponse, ApiTags } from '@nestjs/swagger';
+import { MarkNotificationAsReadDto } from './dtos/mark-notification-as-read.dto';
+
+@ApiBearerAuth()
+@ApiTags('Notification')
+@Controller('notifications')
+export class NotificationsController {
+  constructor(private readonly notificationsService: NotificationsService) {}
+
+  @Patch('/:notificationId/read')
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:12]: it should only be

```suggestion
@Patch('/:notificationId')
```

File: src/modules/notifications/notifications.controller.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  @ApiBody({ type: MarkNotificationAsReadDto, description: 'Read status of the notification' })
+  @ApiResponse({ status: 200, description: 'Notification marked as read successfully' })
+  @ApiResponse({ status: 500, description: 'Internal server error' })
+  @ApiResponse({ status: 400, description: 'Bad request' })
+  @ApiResponse({ status: 400, description: 'Notification not found' })
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:17]: Add a `MarkNotificationAsReadErrorDto` as a type

e.g

```suggestion
@ApiCreatedResponse({ type: CreateNotificationResponseDto, description: 'Notification created successfully' })
  @ApiUnauthorizedResponse({ type: MarkNotificationAsReadErrorDto, description: 'Unauthorized' })
  @ApiBadRequestResponse({ type: MarkNotificationAsReadErrorDto, description: 'Bad Request' })
  @ApiInternalServerErrorResponse({ type: MarkNotificationAsReadErrorDto, description: 'Internal Server Error' })
```

File: src/modules/notifications/notifications.service.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        });
+      }
+
+      const notificationExists = await this.notificationRepository.findOne({
+        where: { id: notification_id },
+        relations: ['user'],
+      });
+
+      if (!notificationExists) {
+        throw new HttpException(
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.service.ts:37]:
```suggestion
throw new NotFoundException{
```

File: src/modules/notifications/notifications.service.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      return {
+        status: 'success',
+        message: 'Notification marked as read successfully',
+        status_code: HttpStatus.OK,
+        data: {
+          notification_id: notificationExists.id,
+          message: notificationExists.message,
+          is_read: notificationExists.isRead,
+          updatedAt: notificationExists.updated_at,
+        },
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.service.ts:59]:
```suggestion
const { id: notification_id, message, isRead: is_read, updated_at} = notificationExists
return {
        status: 'success',
        message: 'Notification marked as read successfully',
        status_code: HttpStatus.OK,
        data: {
          notification_id,
          message,
          is_read,
          updated_at, // change to updated_at and change it on linear
        },
```

File: src/modules/notifications/notifications.service.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      };
+    } catch (error) {
+      if (error instanceof HttpException) {
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.service.ts:62]:
```suggestion
error instanceof HttpException || error instanceof BadRequestException || error instanceof NotFoundException
```

**Review by CFCIfe — COMMENTED** [review]
Review done

File: src/modules/notifications/dtos/create-notification-response.dto.ts
Comment (Heba-WebDev) [inline src/modules/notifications/dtos/create-notification-response.dto.ts:1]: Each class should be in its own file.

Comment (CFCIfe) [reply src/modules/notifications/dtos/create-notification-response.dto.ts:1]: ...to add, validators are not necessary.

File: src/modules/notifications/entities/notifications.entity.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+import { User } from '../../user/entities/user.entity';
+import { Column, Entity, ManyToOne } from 'typeorm';
+
+@Entity()
+export class Notification extends AbstractBaseEntity {
+  @Column({ nullable: false })
+  message: string;
+
+  @Column({ nullable: false, default: false })
+  isRead: boolean;
```
Comment (CFCIfe) [inline src/modules/notifications/entities/notifications.entity.ts:11]: change to `is_read`. Also look for other instances where you use `isRead` and change them

File: src/modules/notifications/entities/notifications.entity.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+  @ManyToOne(() => User, user => user.notifications)
```
Comment (CFCIfe) [inline src/modules/notifications/entities/notifications.entity.ts:13]:
```suggestion
@ManyToOne(() => User, user => user.notifications, { nullable: false, onDelete: 'CASCADE' })
```

File: src/modules/notifications/notifications.controller.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+@ApiBearerAuth()
+@ApiTags('Notification')
+@Controller('notifications')
+export class NotificationsController {
+  constructor(private readonly notificationsService: NotificationsService) {}
+
+  @Patch('/:notificationId')
+  @ApiBody({ type: MarkNotificationAsReadDto, description: 'Read status of the notification' })
+  @ApiCreatedResponse({ type: CreateNotificationResponseDto, description: 'Notification created successfully' })
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:26]:
```suggestion
@ApiOkResponse({
```

**Review by CFCIfe — COMMENTED** [review]
Review comments

**Comment by CFCIfe** [comment]
Please [resolve](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#resolving-conversations) all the conversations

**Review by Heba-WebDev — APPROVED** [review]

**Review by davidolanrewaju — APPROVED** [review]

**Review by Cobidex — APPROVED** [review]
